### PR TITLE
[c++] fix training mismatches between CUDA and CPU

### PR DESCRIPTION
**Pull request by siqi-he**
### Summary

This pr fixes a few bugs uncovered when I was playing with gradient quantization and comparing its results with CPU counterpart. Since each fix is relatively small, instead of spamming with prs, I thought it's probably better to lump them all together. For the fixes here to work, changes introduced in https://github.com/lightgbm-org/LightGBM/pull/7374 need to be applied first. It was purposely left out in this pr to avoid duplication.

#### `798ee9a3`
`exclusive_result = shared_mem_buffer[warpLane - 1];` will be triggered when `threadIdx.x % warpSize == 0`. However this implies `warpLane` will also be 0 so the access will always be out of bound. The `threadIdx.x % warpSize == 0` branch is the first lane of a given warp. Its exclusive value is all earlier warps, i.e. `shared_mem_buffer[warpID - 1]`.

#### `7b00d44c`
Threads own disjoint chunks of the array. The gain loop after the last `GlobalMemoryPrefixSum` reads `hist_*_buffer_ptr[bin]` across chunks with no barrier. The existing `__syncthreads()` sits before the second call, so it orders the gradient chunk writes against those reads but not the hessian ones written during that call. `__syncthreads()` is therefore added at the end of `GlobalMemoryPrefixSum`, which covers every caller and both calls.

#### `477e2fd7`
`left_sum_of_gradients_hessians` and `right_sum_of_gradients_hessians` are defined in `CUDASplitInfo` but are left out in `SplitTreeStructureKernel` and `CUDASplitInfo::operator=`. 

#### `a8adf0cd`
When `!left_is_smaller && (global_thread_index == 14)`, `cuda_hist_pool[right_leaf_index]` is always set to `cuda_hist + 2 * right_leaf_index * num_total_bin`. Under `USE_GRAD_DISCRETIZED` each bin is one packed entry, so the stride should be `num_total_bin` as shown in the `left_is_smaller` branch.

#### `225cdd09`
Non–16-bit quantized bins are packed int64. `FindBestSplitsDiscretizedForLeafKernel` does `reinterpret_cast<const int32_t*>(hist_in_leaf)` and instantiates `BIN_HIST_TYPE = int32_t`, so pointer arithmetic and loads only see half of each bin.

#### `ecc9974c`
`feature_hist_*_buffer` was indexed as `hist_offset * 2` and sized `num_total_bin`. `hist_offset` is already a bin offset, so * 2 is out of bound for features past the halfway point even with a single task. When `MissingType != None`, forward and reverse for the same feature also share that slot, so reverse overwrites forward. The buffer is now sized `2 * num_total_bin` and indexed with `buffer_offset` (hist_offset, plus num_total_bin for reverse). This arguably wastes space as not all features scan in both directions. I feel this is fine here as 1) the space used overall is small 2) it retains the original way of keeping the buffer so less change is required.

#### `c5e89091`
In `FindBestSplitsForLeafKernelInner_GlobalMemory`, reverse `skip_sum` uses `bin >= na_as_missing`. The shared kernel skips `bin < na_as_missing`. Separately, the `bin += blockDim.x` loop overwrites `local_gain` / `threshold_value` with the last valid bin; `ReduceBestGain` needs each thread’s max.

#### `de24ae49`
`FixHistogramKernel` and `FixHistogramDiscretizedKernel` use `threadIdx_x` as the bin index. When `num_bin > blockDim.x`, higher bins never enter the most frequent bin reconstruction. `ShuffleReduceSum` is also clamped to `min(num_bin_aligned, blockDim.x)`, and `__syncthreads()` is added between the full precision grad and hess reductions which share `shared_mem_buffer`. This only triggers at large `max_bin`, which is rare, but since it is not prohibited I figured it is probably best fixed.

#### `6f7b5c92`
`AddPredictionToScoreKernel` remaps `bin = bin - min_bin + offset`, then tests `MissingType::NaN` against column-space max_bin. After remap the last feature bin is `max_bin - min_bin + offset`.

#### `11e3db47`
When gradient quantization is not used, `FindBestSplitsForLeafKernelInner` recovers omitted bin 0 when `na_as_missing && mfb_offset == 1`. `FindBestSplitsDiscretizedForLeafKernelInner` has no such path, so implicit bin-0 mass is missing from the quantized prefix sum. The changes are mostly adapted from the code path without quantization. 

### Validation

Attached is a script I used to show the bugs and confirm the fixes worked. Instructions on how to run and test are included. As a reminder, changes in https://github.com/lightgbm-org/LightGBM/pull/7374 need to be incorporated first. While the fixes in `798ee9a3` and `7b00d44c` are unit-testable, there are no existing C++ tests for the CUDA path, so I omitted adding those tests here. The script has six cases for the eight remaining fixes: `477e2fd7`+`a8adf0cd` share `quantized_multi_leaf`, and `ecc9974c`+`c5e89091` share `global_nan`; those pairs cannot be separated end-to-end.

<details>
  <summary>full test script</summary>

```python
#!/usr/bin/env python
"""CUDA quantization / correctness checks.

Uncomment cases from top to bottom as matching fixes are applied.
Fail on unfixed code; pass after the fix.

    python run.py
"""

from __future__ import annotations

import os
from pathlib import Path

import numpy as np
import pandas as pd
from sklearn.metrics import log_loss, roc_auc_score
from sklearn.preprocessing import scale

HERE = Path(__file__).resolve().parent
SEED = 42
QUANT_BINS = 4
# Bosch dataset is downloaded from kaggle: https://www.kaggle.com/c/bosch-production-line-performance
BOSCH_CSV = Path(os.environ.get("BOSCH_CSV", HERE.parent / "train_numeric.csv"))


def synth(rows: int, feats: int, zeros: float = 0.0, nans: float = 0.0, seed: int = SEED):
    """Synthetic binary data; optional zero/NaN fractions are also predictive of y."""
    rng = np.random.RandomState(seed)
    if zeros:
        x = (np.abs(rng.normal(5.0, 2.0, size=(rows, feats))) + 1.0).astype(np.float32)
    else:
        x = rng.normal(size=(rows, feats)).astype(np.float32)
    u = rng.uniform(size=x.shape)
    zero_mask = (u < zeros) if zeros else None
    if zero_mask is not None:
        x[zero_mask] = 0.0
    nan_mask = ((u >= zeros) & (u < zeros + nans)) if nans else None

    logits = 1.2 * scale(np.nan_to_num(x, nan=0.0) @ rng.normal(size=feats))
    src = zero_mask if (zeros and nans) else nan_mask
    if src is not None:
        logits = logits + 0.8 * scale(src[:, : max(1, feats // 2)].sum(axis=1))
    y = (rng.uniform(size=rows) < 1 / (1 + np.exp(-logits))).astype(np.float32)
    if nan_mask is not None:
        x[nan_mask] = np.nan
    return x, y


def load_bosch(rows: int = 80_000, n_features: int = 120):
    if not BOSCH_CSV.exists():
        raise FileNotFoundError(
            f"missing Bosch CSV: {BOSCH_CSV}\n"
            "place train_numeric.csv under higgs_training/ or set BOSCH_CSV="
        )
    header = pd.read_csv(BOSCH_CSV, nrows=0)
    feature_cols = [c for c in header.columns if c not in ("Id", "Response")][:n_features]
    df = pd.read_csv(BOSCH_CSV, usecols=feature_cols + ["Response"], nrows=rows,
                     dtype=np.float32)
    return df[feature_cols].to_numpy(), df["Response"].to_numpy()


def train(x, y, device, max_bin, quantized, num_leaves, rounds,
          min_data_in_leaf=20, internal_metric=False):
    import lightgbm as lgb

    split = int(0.8 * x.shape[0])
    params = {
        "objective": "binary",
        "metric": "binary_logloss" if internal_metric else "None",
        "device_type": device, "gpu_device_id": 0,
        "learning_rate": 0.1, "num_leaves": num_leaves, "max_bin": max_bin,
        "use_quantized_grad": quantized, "num_grad_quant_bins": QUANT_BINS,
        "min_data_in_leaf": min_data_in_leaf, "num_threads": 8,
        "seed": SEED, "deterministic": True, "verbosity": -1,
    }
    ds_params = {"max_bin": max_bin}
    train_set = lgb.Dataset(x[:split], label=y[:split], params=ds_params)
    result = {}
    if internal_metric:
        valid = lgb.Dataset(x[split:], label=y[split:], reference=train_set, params=ds_params)
        booster = lgb.train(params, train_set, num_boost_round=rounds, valid_sets=[valid],
                            valid_names=["v"], callbacks=[lgb.record_evaluation(result)])
    else:
        booster = lgb.train(params, train_set, num_boost_round=rounds)
    pred = booster.predict(x[split:])
    return {
        "auc": float(roc_auc_score(y[split:], pred)),
        "logloss": float(log_loss(y[split:], pred)),
        "internal_logloss": float(result["v"]["binary_logloss"][-1]) if internal_metric else None,
    }


def cpu_vs_cuda(name, x, y, max_bin, quantized, num_leaves, rounds,
                min_data_in_leaf=20):
    """Shared bounds with test_cuda.py: |ΔAUC| < 0.01 and |logloss ratio − 1| < 0.02 (needs #7374 for quantized)."""
    kw = dict(max_bin=max_bin, quantized=quantized, num_leaves=num_leaves,
              rounds=rounds, min_data_in_leaf=min_data_in_leaf)
    cpu = train(x, y, device="cpu", **kw)
    cuda = train(x, y, device="cuda", **kw)
    auc_delta = cuda["auc"] - cpu["auc"]
    ratio = cuda["logloss"] / max(cpu["logloss"], 1e-12)
    ok = abs(auc_delta) < 0.01 and abs(ratio - 1.0) < 0.02
    print(f"{'PASS' if ok else 'FAIL'}  {name}  "
          f"auc_delta={auc_delta:+.4f}  logloss_ratio={ratio:.4f}")
    assert ok, name


def bosch_score(name="bosch_score"):
    x, y = load_bosch(80_000, n_features=120)
    r = train(x, y, device="cuda", max_bin=255, quantized=False,
              num_leaves=15, rounds=20, internal_metric=True)
    ratio = r["internal_logloss"] / max(r["logloss"], 1e-12)
    ok = abs(ratio - 1.0) < 0.5
    print(f"{'PASS' if ok else 'FAIL'}  {name}  "
          f"internal_over_predict={ratio:.4f}  internal_logloss={r['internal_logloss']:.4f}")
    np.testing.assert_allclose(r["internal_logloss"], r["logloss"], rtol=0.5)


if __name__ == "__main__":
    # Uncomment top → bottom as matching fixes are applied on
    # gpu-quantization-bug-fixes (798ee9a3+7b00d44c, then 477e2fd7, a8adf0cd,
    # 225cdd09, ecc9974c, c5e89091, de24ae49, 6f7b5c92, 11e3db47).
    # Baseline needs PR #7374.

    # quantized_multi_leaf  (477e2fd7+a8adf0cd)
    cpu_vs_cuda("quantized_multi_leaf", *synth(15_000, 20),
                max_bin=255, quantized=True, num_leaves=31, rounds=25)

    # quantized_32bit  (225cdd09)
    # cpu_vs_cuda("quantized_32bit", *synth(50_000, 20),
    #             max_bin=255, quantized=True, num_leaves=2, rounds=25)

    # global_nan  (ecc9974c + c5e89091) — covers both; a global-path NaN test
    # necessarily exercises the REVERSE skip predicate and the strided
    # per-thread best-bin loop, so the two cannot be separated end-to-end.
    # cpu_vs_cuda("global_nan", *synth(60_000, 32, nans=0.25),
    #             max_bin=300, quantized=False, num_leaves=15, rounds=30)

    # fixhistogram_high_bins (de24ae49) — after 798ee9a3+7b00d44c and through c5e89091;
    # dense features so most_freq_bin != 0 and FixHistogram actually runs.
    # cpu_vs_cuda("fixhistogram_high_bins", *synth(80_000, 16),
    #             max_bin=768, quantized=False, num_leaves=2, rounds=30,
    #             min_data_in_leaf=100)

    # bosch_score  (6f7b5c92)
    # bosch_score()

    # quantized_nan_zero  (11e3db47)
    # cpu_vs_cuda("quantized_nan_zero", *synth(50_000, 40, zeros=0.60, nans=0.20),
    #             max_bin=255, quantized=True, num_leaves=31, rounds=25,
    #             min_data_in_leaf=50)

```
</details>

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for your interest in LightGBM. We have a long backlog of reviews to catch up on so it may be a while until I can look. And like you alluded to, there are other in-progress fixes to the CUDA build we'd want to get in first.

one note at a glance

> _Attached is a script I used to show the bugs and confirm the fixes worked. Instructions on how to run and test are included. _

Can you please check in some Python unit tests that'd expose these bugs and prevent us from reintroducing them? Follow the existing patterns you see in `tests/python_package_test/`.

**Comment by siqi-he** [comment]
@jameslamb no worries. I've added tests as suggested. note that as it is tests with gradient quantization will fail due to the bug identified in #7374. once it's checked in, the tests should clear.
